File: py-langgraph/src/api/routes/stream.py
import asyncio
from uuid import uuid4

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from src.api.auth import get_agent_tool_identity, require_agent_user_id
from src.api.request_logging import log_request_error
from src.api.sse import SseEventSequencer, encode_sse_heartbeat, with_sse_heartbeats
from src.services import (
    AgentService,
    BusinessError,
    ConversationService,
    Message,
    TurnService,
    wait_for_conversation_persistence,
)

router = APIRouter()


class StreamRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=16_000)
    thread_id: str | None = None
    user_id: str | None = None
    client_message_id: str | None = Field(default=None, min_length=1, max_length=128)


# 流式对话接口 — 使用 Server-Sent Events 逐字返回 AI 回复
@router.post("")
# 执行 stream 对应的业务逻辑
async def stream(payload: StreamRequest, request: Request):
    """流式对话 — 使用 Server-Sent Events"""
    from fastapi.responses import StreamingResponse

    trusted_user_id = require_agent_user_id(request, payload.user_id)
    tool_identity = get_agent_tool_identity(request, payload.user_id)
    thread_id = payload.thread_id or str(uuid4())

    # 确保会话存在于当前仓储，并校验传入 thread_id 的用户归属。
    ConversationService.ensure(thread_id, trusted_user_id)
    turn, created = TurnService.begin(
        thread_id, trusted_user_id, payload.client_message_id
    )
    completed = None if created else TurnService.completed_message(turn)
    if not created and not completed:
        error = TurnService.duplicate_error(turn["status"])
        raise HTTPException(status_code=error.status_code, detail=error.to_dict())
    if created:
        await wait_for_conversation_persistence(thread_id)
        user_message = ConversationService.append_user_message(
            thread_id, payload.message, trusted_user_id
        )
        TurnService.start(turn["id"], trusted_user_id, user_message.id)

    # 生成 SSE 事件流，复用 V1 的统一 AgentService 编排
    async def event_generator():
        sse_events = SseEventSequencer(turn["id"])
        full_answer = ""
        turn_completed = completed is not None
        yield sse_events.event(
            "meta", {"conversation_id": thread_id, "thread_id": thread_id, "turn_id": turn["id"]}
        )
        try:
            if completed:
                yield sse_events.event(
                    "text",
                    {"delta": completed.content, "text": completed.content, "partial": False, "replayed": True},
                )
                yield sse_events.done()
                return
            events = AgentService.chat_stream(
                thread_id, payload.message, user_id=trusted_user_id, tool_identity=tool_identity
            )
            async for event in with_sse_heartbeats(events):
                if await request.is_disconnected():
                    TurnService.terminate(turn["id"], trusted_user_id, True, "CLIENT_CANCELLED")
                    return
                if event is None:
                    yield encode_sse_heartbeat()
                    continue
                if event["type"] == "text":
                    full_answer += event["text"]
                    response_payload = {
                        "delta": event["text"],
                        "text": event["text"],
                        "partial": event.get("partial", False),
                    }
                    event_name = "text"
                elif event["type"] == "tool":
                    response_payload = {
                        "event": "tool",
                        "tool_name": event["tool_name"],
                        "status": event["status"],
                        "call_id": event["call_id"],
                        "duration_ms": event.get("duration_ms"),
                    }
                    event_name = "tool"
                else:
                    response_payload = {
                        "state": event.get("state"),
                        "stop_reason": event.get("stop_reason"),
                        "react": event.get("react"),
                    }
                    event_name = event.get("event", "agent")
                yield sse_events.event(
                    event_name,
                    response_payload,
                )
            persisted = next(
                (
                    item
                    for item in reversed(ConversationService.get_messages(thread_id))
                    if item["role"] == "assistant"
                ),
                None,
            )
            assistant = (
                Message("assistant", persisted["content"], persisted["id"], persisted["created_at"])
                if persisted
                else Message("assistant", full_answer)
            )
            TurnService.complete(turn["id"], trusted_user_id, assistant)
            turn_completed = True
        except BusinessError as error:
            log_request_error(
                request,
                error,
                {"thread_id": thread_id, "user_id": trusted_user_id, "stream": True},
            )
            yield sse_events.event(
                "error", {"ok": False, "error": error.to_dict().get("error", error.to_dict())}
            )
            if not turn_completed:
                try:
                    TurnService.terminate(turn["id"], trusted_user_id, False, error.code.value)
                except Exception:
                    pass
        except asyncio.CancelledError:
            if not turn_completed:
                try:
                    TurnService.terminate(turn["id"], trusted_user_id, True, "CLIENT_CANCELLED")
                except Exception:
                    pass
            log_request_error(
                request,
                RuntimeError("Client disconnected"),
                {"thread_id": thread_id, "user_id": trusted_user_id, "stream": True},
            )
            raise
        except Exception as error:
            log_request_error(
                request, error, {"thread_id": thread_id, "user_id": trusted_user_id, "stream": True}
            )
            yield sse_events.event(
                "error",
                {"ok": False, "error": {"code": "INTERNAL_ERROR", "message": "处理请求时发生错误"}},
            )
            if not turn_completed:
                try:
                    TurnService.terminate(turn["id"], trusted_user_id, False, "INTERNAL_ERROR")
                except Exception:
                    pass
        yield sse_events.done()

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache, no-transform", "X-Accel-Buffering": "no"},
    )

File: ts-langchain/src/api/routes/stream.ts
import type { FastifyInstance } from "fastify";
import {
  AgentService,
  ConversationService,
  waitForConversationPersistence,
  TurnService,
} from "../../services/index.js";
import { getAgentToolIdentity, requireAgentUserId } from "../middleware/auth.js";
import { logRequestError } from "../middleware/error.js";
import { SseEventSequencer, encodeSseHeartbeat } from "../sse.js";

// 向 SSE 客户端写入一条事件，并在内核背压时等待 drain。
async function writeSse(raw: any, data: string): Promise<void> {
  if (raw.writableEnded || raw.destroyed) return;
  if (!raw.write(data)) {
    await new Promise<void>((resolve) => {
      const finish = () => {
        raw.removeListener("drain", finish);
        raw.removeListener("close", finish);
        resolve();
      };
      raw.once("drain", finish);
      raw.once("close", finish);
    });
  }
}

// 创建或注册 registerStreamRoutes 所需的数据
export async function registerStreamRoutes(app: FastifyInstance) {
  app.post<{ Body: { message: string; thread_id?: string; user_id?: string; client_message_id?: string } }>(
    "/stream",
    async (request, reply) => {
      const { message, thread_id, user_id, client_message_id } = request.body;
      const trustedUserId = requireAgentUserId(request, user_id);
      const toolIdentity = getAgentToolIdentity(request, user_id);
      if (typeof message !== "string" || message.length < 1 || message.length > 16_000) {
        return reply.status(400).send({
          error: { code: "VALIDATION_ERROR", message: "请求参数无效" },
        });
      }

      const threadId = thread_id || crypto.randomUUID();

      // 确保会话存在于当前仓储，并校验传入 thread_id 的用户归属。
      await ConversationService.ensure(threadId, trustedUserId);
      const turnResult = await TurnService.begin(threadId, trustedUserId, client_message_id);
      const completed = turnResult.created ? null : TurnService.completedMessage(turnResult.turn);
      if (!turnResult.created && !completed) {
        const error = TurnService.duplicateError(turnResult.turn.status);
        return reply.status(error.statusCode).send(error.toJSON());
      }
      if (turnResult.created) {
        await waitForConversationPersistence(threadId);
        const userMessage = await ConversationService.appendUserMessage(threadId, message);
        await TurnService.start(turnResult.turn.id, trustedUserId, userMessage.id);
      }

      reply.raw.setHeader("Content-Type", "text/event-stream");
      reply.raw.setHeader("Cache-Control", "no-cache, no-transform");
      reply.raw.setHeader("Connection", "keep-alive");
      reply.raw.setHeader("X-Accel-Buffering", "no");
      reply.raw.flushHeaders();
      const requestController = new AbortController();
      const abortOnDisconnect = () => {
        if (!reply.raw.writableEnded) {
          requestController.abort(new Error("Client disconnected"));
        }
      };
      request.raw.once("aborted", abortOnDisconnect);
      reply.raw.once("close", abortOnDisconnect);
      const heartbeat = setInterval(() => {
        void writeSse(reply.raw, encodeSseHeartbeat());
      }, 15_000);
      heartbeat.unref();
      const sseEvents = new SseEventSequencer(turnResult.turn.id);
      let fullAnswer = "";
      let turnCompleted = Boolean(completed);

      try {
        await writeSse(reply.raw, sseEvents.event("meta", {
          conversation_id: threadId,
          thread_id: threadId,
          turn_id: turnResult.turn.id,
        }));
        if (completed) {
          await writeSse(reply.raw, sseEvents.event("text", { delta: completed.content, text: completed.content, partial: false, replayed: true }));
          return;
        }
        for await (const event of AgentService.chatStream(
          threadId,
          message,
          trustedUserId,
          requestController.signal,
          toolIdentity,
        )) {
          if (event.type === "text") {
            fullAnswer += event.text;
            await writeSse(
              reply.raw,
              sseEvents.event("text", {
                delta: event.text,
                text: event.text,
                partial: event.partial ?? false,
              }),
            );
          } else if (event.type === "tool") {
            await writeSse(
              reply.raw,
              sseEvents.event("tool", {
                event: "tool",
                tool_name: event.toolName,
                status: event.status,
                call_id: event.callId,
                duration_ms: event.durationMs ?? null,
              }),
            );
          } else {
            await writeSse(
              reply.raw,
              sseEvents.event(event.event, {
                state: event.state,
                stop_reason: event.stopReason,
                react: event.react,
              }),
            );
          }
        }
        const assistantMessage = (await ConversationService.getMessages(threadId)).reverse().find((item) => item.role === "assistant")
          ?? { id: crypto.randomUUID(), role: "assistant" as const, content: fullAnswer, createdAt: new Date().toISOString() };
        await TurnService.complete(turnResult.turn.id, trustedUserId, assistantMessage);
        turnCompleted = true;
      } catch (error: any) {
        logRequestError(request, error, {
          thread_id: threadId,
          user_id: trustedUserId,
          stream: true,
        });
        await writeSse(
          reply.raw,
          sseEvents.event("error", {
            ok: false,
            error: {
              code: error?.code || "INTERNAL_ERROR",
              message: error?.message || "处理请求时发生错误",
            },
          }),
        );
        if (!turnCompleted) {
          await TurnService.terminate(
            turnResult.turn.id,
            trustedUserId,
            requestController.signal.aborted,
            requestController.signal.aborted ? "CLIENT_CANCELLED" : error?.code || "INTERNAL_ERROR",
          ).catch(() => undefined);
        }
      } finally {
        clearInterval(heartbeat);
        request.raw.removeListener("aborted", abortOnDisconnect);
        reply.raw.removeListener("close", abortOnDisconnect);
        try {
          await writeSse(reply.raw, sseEvents.done());
        } finally {
          reply.raw.end();
        }
      }
    },
  );
}
